kani-driver/src/assess/table_builder.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains code that helps build tables more easily.
+//!
+//! Virtually all tables are we build via the same method:
+//!
+//! 1. There is some input data with irregular structure
+//! 2. There is an implied (we avoid writing it) type with regular structure
+//! 3. There is the type of the table row
+//!
+//! As an abstract example, #1 is often something like `[(Node, [Node])]` which you can think of
+//! as a graph. #2 is a normalized version like `[(Node, Node)]` which is an ordinary
+//! adjacency list. #3 ends up being a structure like `(Node name, outgoing edge count)`.
+//!
+//! With this module, you construct a table by:
+//!
+//! 1. Create a type for a table row (#3 above).
+//! 2. Write the code to traverse your irregular structure (#1) and produce table rows.
+//! 3. Implement [`TableRow`] for your table row type, which implements
+//!    an [`TableRow::merge`] method. This avoids having to create a type for
+//!    the intermediate "regular" structure (#2 above), by instead thinking in
+//!    terms of merging rows by their [`TableRow::key`]. (Vaguely like map-reduce.)
+//! 4. Use [`TableBuilder`] to construct the table.
+//! 5. [Optional] Implement [`RenderableTableRow`] so you can print the table.
+
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use comfy_table::Table;
+
+/// A [`TableRow`] is a type where a multiple elements with the same `key` should be `merge`d into one.
+/// This type is used in conjuction with [`TableBuilder`].
+///
+/// We further give an ordering (via `compare`) that defines how rows should appear in the final table.
+pub trait TableRow {
+    type Key: Eq + Hash;
+
+    /// Returns the key for this row entry, so that multiple entries with the same key can be aggregated
+    fn key(&self) -> Self::Key;
+    /// Merges two rows into one. `self` is always the existing (or first) entry, `new` is what should be merged in.
+    fn merge(&mut self, new: Self);
+    /// Define the order rows should appear in a table. Not needed for merging, but required for producing results.
+    fn compare(&self, right: &Self) -> Ordering;
+}
+
+/// What kind of data appears in a table column, when visually rendered for the user.
+pub enum ColumnType {
+    /// Text might be very wide and need limiting, and it should be left-justified.
+    Text,
+    /// Numbers should be right-justified.
+    Number,
+}
+
+/// Describes how a table should be printed visually.
+///
+/// This is an extension to [`TableRow`] (which is solely concerned with computing a table's contents.)
+/// In this trait, we are solely concerned with how the table should look when printed.
+pub trait RenderableTableRow {
```

Comment:
I would prefer if we return a single vector of tuples or something similar. Having different methods return a vector that should be of the same size for all of them makes it a bit error prone.

Reply:
Unfortunately it complicates code a little to aggregate them, just due to how the table rendering library works.

I did add test cases checking length, and as a bonus I realized with just `#[derive(Default)]` I could also check the length of the `row` result, too.

(I also thought about some more advanced type magic to return fixed-size arrays instead of vectors, but I felt that would be complicating this code too much.)